### PR TITLE
PWN-7169 - state handlers, new states

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7169

## Description of Work

state handlers, new states

### CODE REVIEW

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 1]
```diff
@@ -3,31 +3,93 @@ package org.p2p.wallet.swap.jupiter.statemanager
...
+import kotlin.coroutines.cancellation.CancellationException
 import kotlinx.coroutines.Job
 import kotlinx.coroutines.SupervisorJob
 import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.StateFlow
 import kotlinx.coroutines.launch
 
+private const val DELAY_IN_MILLIS = 20_000L
```
[the remark this answers is not in the document]
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:17]: Снимок экрана 2023-02-22 в 13 24 13
not sure

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 2]
```diff
@@ -3,31 +3,93 @@ package org.p2p.wallet.swap.jupiter.statemanager
...
+                }
+            } catch (cancelled: CancellationException) {
+                Timber.i(cancelled)
+            } catch (featureException: FeatureException) {
+                Timber.e(featureException)
+                state.value = SwapState.SwapException.FeatureExceptionWrapper(
+                    previousFeatureState = actualNoErrorState(),
+                    featureException = featureException,
+                )
+            } catch (exception: Exception) {
```
[the remark this answers is not in the document]
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:57]: fix in common brancefeature/jupiter-swap-all
next pr

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 3]
```diff
@@ -3,31 +3,93 @@ package org.p2p.wallet.swap.jupiter.statemanager
...
+    private fun startRefreshJob() {
+        refreshJob = launch {
+            try {
+                while (refreshJob?.isActive == true) {
+                    delay(DELAY_IN_MILLIS)
+                    handleNewAction(SwapStateAction.RefreshRoutes)
+                }
+            } catch (e: Exception) {
+                Timber.e(e)
+                // todo ignore?
```
[the remark this answers is not in the document]
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:82]: next pr

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:17]: Screen Shot 2023-02-22 at 17 32 47
the import is wrong here

[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:17]: should be import kotlin.time.Duration.Companion.seconds

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt [hunk 4]
```diff
@@ -3,31 +3,93 @@ package org.p2p.wallet.swap.jupiter.statemanager
...
+            }
+        }
+    }
+
+    private fun actualNoErrorState(): SwapState {
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateManager.kt:87]: strange thing =(
state swapped if was SwapException

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        action: SwapStateAction
+    ) {
+        val oldState = state as SwapState.InitialLoading
+
+        when (action) {
+            is SwapStateAction.InitialLoading -> {
+                stateFlow.value = SwapState.InitialLoading
+
+                val jupiterTokens = jupiterTokensRepository.getTokens()
+                val first = jupiterTokens.find { it.tokenName == "SOL" } ?: return
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt:28]: Constants.SOL_SYMBOL

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt [hunk 6]
```diff
@@ -81,7 +81,7 @@ class SwapWidgetMapper {
     private fun fiatAmount(token: SwapTokenModel, tokenAmount: BigDecimal?): TextViewCellModel.Raw? {
         if (tokenAmount == null) return null
         val ratio = when (token) {
-            is SwapTokenModel.JupiterToken -> token.details.priceInUsd
+            is SwapTokenModel.JupiterToken -> /*token.details.priceInUsd*/null // todo
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:84]: // todo ticket number

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt (hunk 5 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt:28]: or better - compare by mints

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt:28]: first -> tokenA
second -> tokenB

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt (hunk 5 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/handler/SwapStateInitialLoadingHandler.kt:28]: it's for testing, remove here https://github.com/p2p-org/key-app-android/pull/1290

**Pull request merged**